river/Vector.zig

```diff
@@ -0,0 +1,60 @@
+// This file is part of river, a dynamic tiling wayland compositor.
+//
+// Copyright 2023 The River Developers
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+const std = @import("std");
+const math = std.math;
+const wlr = @import("wlroots");
+
+const Vector = @This();
+
+x: i32,
+y: i32,
+
+pub fn positionOfBox(box: wlr.Box) Vector {
+    return .{
+        .x = box.x + @divFloor(box.width, 2),
+        .y = box.y + @divFloor(box.height, 2),
+    };
+}
+
+/// Returns the difference between two vectors.
+pub fn diff(a: Vector, b: Vector) Vector {
+    return .{
+        .x = b.x - a.x,
+        .y = b.y - a.y,
+    };
+}
+
+/// Returns the direction of the vector.
+pub fn direction(self: Vector) ?wlr.OutputLayout.Direction {
+    // A zero length vector has no direction
+    if (self.x == 0 and self.y == 0) return null;
+
+    if ((math.absInt(self.y) catch return null) > (math.absInt(self.x) catch return null)) {
+        // Careful: We are operating in a Y-inverted coordinate system.
+        return if (self.y > 0) .down else .up;
+    } else {
+        return if (self.x > 0) .right else .left;
+    }
+}
+
+/// Returns the length of the vector.
+pub fn length(self: Vector) u31 {
+    const x = @intCast(u31, math.absInt(self.x) catch math.maxInt(i32));
+    const y = @intCast(u31, math.absInt(self.y) catch math.maxInt(i32));
+    return math.sqrt((x *| x) +| (y *| y));
```

Comment:
The first 2 lines here are unnecessary I believe, `self.x *| self.x` will always be positive and handles the case of the minimum integer value correctly.

Reply:
```
/usr/lib/zig/std/math/sqrt.zig:29:24: error: sqrt not implemented for signed integers 
            .signed => @compileError("sqrt not implemented for signed integers"), 
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

Reply:
Mathematically you are right, but the typesystem disagrees.

Reply:
Oh, unless you mean it with still having an intCast inside the sqrt, that works.

Reply:
Ah, I didn't realize `sqrt` doesn't support signed integers. Doing the `@intCast()` after the multiplication is cleaner though, thanks!